server/src/main/java/org/elasticsearch/index/mapper/FallbackSyntheticSourceBlockLoader.java

```diff
@@ -70,46 +85,73 @@ public SortedSetDocValues ordinals(LeafReaderContext context) throws IOException
         throw new UnsupportedOperationException();
     }
 
+    public static Set<String> splitIntoFieldPaths(String fieldName) {
+        var paths = new HashSet<String>();
+        paths.add("_doc");
+        var current = new StringBuilder();
+        for (var part : fieldName.split("\\.")) {
+            if (current.isEmpty() == false) {
+                current.append('.');
+            }
+            current.append(part);
+            paths.add(current.toString());
+        }
+        return paths;
+    }
+
     private static class IgnoredSourceRowStrideReader<T> implements RowStrideReader {
         // Contains name of the field and all its parents
-        private final Set<String> fieldNames;
         private final String fieldName;
+        private final Set<String> fieldPaths;
         private final Reader<T> reader;
+        private final IndexVersion indexCreatedVersion;
 
-        IgnoredSourceRowStrideReader(String fieldName, Reader<T> reader) {
+        IgnoredSourceRowStrideReader(String fieldName, Set<String> fieldPaths, Reader<T> reader, IndexVersion indexCreatedVersion) {
             this.fieldName = fieldName;
+            this.fieldPaths = fieldPaths;
             this.reader = reader;
-            this.fieldNames = new HashSet<>() {
-                {
-                    add("_doc");
-                }
-            };
-
-            var current = new StringBuilder();
-            for (String part : fieldName.split("\\.")) {
-                if (current.isEmpty() == false) {
-                    current.append('.');
-                }
-                current.append(part);
-                fieldNames.add(current.toString());
-            }
-
+            this.indexCreatedVersion = indexCreatedVersion;
         }
 
         @Override
         public void read(int docId, StoredFields storedFields, Builder builder) throws IOException {
-            var ignoredSource = storedFields.storedFields().get(IgnoredSourceFieldMapper.NAME);
-            if (ignoredSource == null) {
-                builder.appendNull();
-                return;
-            }
-
             Map<String, List<IgnoredSourceFieldMapper.NameValue>> valuesForFieldAndParents = new HashMap<>();
 
-            for (Object value : ignoredSource) {
-                IgnoredSourceFieldMapper.NameValue nameValue = IgnoredSourceFieldMapper.decode(value);
-                if (fieldNames.contains(nameValue.name())) {
-                    valuesForFieldAndParents.computeIfAbsent(nameValue.name(), k -> new ArrayList<>()).add(nameValue);
+            if (indexCreatedVersion.onOrAfter(IndexVersions.IGNORED_SOURCE_FIELDS_PER_ENTRY)) {
```

Comment:
Maybe it makes sense to have two `IgnoredSourceRowStrideReader` implementations? Something like `LegacySingleIgnoredSourceRowStrideReader` and `PerFieldIgnoredSourceRowStrideReader`?